### PR TITLE
llama: fix memory leak for grammar

### PR DESCRIPTION
**Pull request by ParthSareen**
Closes https://github.com/ollama/ollama/issues/10688

### CODE REVIEW

**Review by jmorganca — APPROVED** [review]
LGTM once tests pass

File: llama/sampling_ext.cpp [hunk 1]
```diff
@@ -114,8 +114,11 @@ void grammar_free(struct llama_grammar *g) {
         if (g->vocab != nullptr) {
             delete g->vocab;
         }
-        llama_grammar_free_impl(g);
     }
+    if (g->o_vocab != nullptr) {
```
Comment (rick-github) [inline llama/sampling_ext.cpp:118]: This assumes g is never nil.  If that's true, does it need to be tested in the statement further up?  Or should the nil test be done here as well?

Comment (ParthSareen) [reply llama/sampling_ext.cpp:118]: _Should be okay_ - `llama.go` checks the nil before calling the free:
```go
func (g *Grammar) Free() {
	g.mu.Lock()
	defer g.mu.Unlock()
	if g.c != nil {
		C.grammar_free(g.c)
		g.c = nil
	}
}
```

Comment (rick-github) [reply llama/sampling_ext.cpp:118]: So the check on line 113 is not required.

Comment (jmorganca) [reply llama/sampling_ext.cpp:118]: @ParthSareen yes this should be in the `if (g != nullptr) {` block

Comment (ParthSareen) [reply llama/sampling_ext.cpp]: Fixed!

**Review by jmorganca — APPROVED** [review]
